### PR TITLE
[PUBLISHING] [CONTENT_API] [NINJS]: added "firstpublished" field

### PR DESCRIPTION
**Pull request by jerome-poisson**
"firstpublished" is set to current datetime on first publication, then
it is kept unchanged.
NINJS schema extension has been updated accordingly

SDFID-175

### CODE REVIEW

File: content_api/publish/service.py [hunk 1]
```diff
@@ -38,10 +38,18 @@ def publish(self, item, subscribers=[]):
 
         :param item: item to publish
         """
-
         if not self._filter_item(item):
-            doc = self.formatter._transform_to_ninjs(item, self.subscriber)
             now = utcnow()
+            try:
+                # do we already have an item id db?
+                stored_item = self.find_one(req=None, _id=item['_id'])
```
Comment (petrjasek) [inline content_api/publish/service.py]: would rather put this into _create_doc which is already checking previous version so it's not fetched twice

File: apps/publish/enqueue/__init__.py [hunk 2]
```diff
@@ -95,8 +95,11 @@ def enqueue_item(published_item):
             # if scheduled then change the state to published
             # change the `version` and `versioncreated` for the item
             # in archive collection and published collection.
-            versioncreated = utcnow()
+            now = utcnow()
+            versioncreated = now
```
Comment (mugurrus) [inline apps/publish/enqueue/__init__.py]: You can put these two in one line

**Comment by jerome-poisson** [comment]
@petrjasek need to get `firstpublished` before generating ninjs. So I still get the item at the same place, but not I transmit it to _create_doc so it's not fetched 2 times.

**Comment by petrjasek** [comment]
why is firstpublished needed before doing ninjs?

**Comment by jerome-poisson** [comment]
@petrjasek because it is required in ninjs, that's what is requested in the issue (the schema is updated by this patch)

**Comment by jerome-poisson** [comment]
@petrjasek Hi, waiting for your green light before merging, let me know if you think it should be modified, thanks.

**Comment by petrjasek** [comment]
I don't see it in the requirements, it says it should be there when you fetch something from content api, and it says it's missing in ninjs, but not that it should be added there

File: content_api/publish/service.py [hunk 3]
```diff
@@ -38,10 +38,20 @@ def publish(self, item, subscribers=[]):
...
             now = utcnow()
+            try:
+                # do we already have an item id db?
+                original_item = self.find_one(req=None, _id=item.get('guid') or item['_id'])
+            except KeyError:
+                original_item = None
+                check_original = True
+            else:
+                # if yes, we need to use previous firstpublished data
+                item.setdefault('firstpublished',
```
Comment (petrjasek) [inline content_api/publish/service.py]: I'm still not sure why is this needed, it should be set on item when publishing and corrections shouldn't update it anyhow, so what's the usecase here?

Comment (petrjasek) [reply content_api/publish/service.py]: imo it should be set when user clicks on publish, before it enters the output handling, and not when it's pushed to contentapi. so if you get an item from content api or via ftp or whatever there will be the same timestamp.

Comment (jerome-poisson) [reply content_api/publish/service.py]: OK I see, I'll move it.

**Comment by jerome-poisson** [comment]
checked with @petrjasek it's alright now. As soon as all is green we can merge this patch.
